### ISSUE TITLE
I keep ending up with a 0 byte executable

### ISSUE THREAD

**Issue by Nickhoyer** [jaemk/self_update#81]
I tried changing my main function from async to sync, I tried running as admin, running in a different folder, closing the app after updating, but every time, it looks to be correctly downloading a 13.2MB file from github, but then without warnings, it replaces my local exe with a 0 byte file that will never run.

Is there anything I can do to avoid this? I didn't see it mentioned in other issues. I'm on windows, and so is my ci runner.
![image](https://user-images.githubusercontent.com/9862931/149938481-e107b714-1cce-4d96-91ff-e723698d4721.png)

**Comment by jaemk** [comment]
hey @Nickhoyer , could you try using some variation of the code in the docs here https://github.com/jaemk/self_update/blob/cf0a017f058bbeaf63ad680ff8c5a60513efad8c/src/lib.rs#L92 and tell it to move the downloaded/extracted file to some location that isn't the current exe? I'm wondering if it's something related to exe replacement on windows.

**Comment by andrew121410** [comment]
Same, I keep getting 0 bytes. Not sure how to fix it.

**Comment by andrew121410** [comment]
Looking at the photo [Nickhoyer](https://github.com/Nickhoyer) posted it's not a download link for a executable, but a page with JSON on it. If that makes sense?

For an example here's mine

![Screenshot 2022-04-19 173419](https://user-images.githubusercontent.com/12376951/164105709-d161e270-7851-4ba7-8201-35acac1d15d8.png)

Going to the "new exe download url" shows

![Screenshot 2022-04-19 173603](https://user-images.githubusercontent.com/12376951/164105961-2112412d-8bba-404e-af42-94211e09b968.png)

So maybe this is the problem? I have no idea lol

**Comment by jaemk** [comment]
The github api returns json to basic get requests. You have to set `Accept: application/octet-stream` to get back the contents of the release (which happens [here](https://github.com/jaemk/self_update/blob/3d522e87d714c41f77c33cc50573567546da1774/src/update.rs#L217])):
```
james@tamale|t.main:~/bin$
::: curl -LO https://api.github.com/repos/andrew121410/Limonium/releases/assets/63040612
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1561  100  1561    0     0  10518      0 --:--:-- --:--:-- --:--:-- 11150
james@tamale|t.main:~/bin$
::: file 63040612
63040612: JSON data

::: curl -LO https://api.github.com/repos/andrew121410/Limonium/releases/assets/63040612 -H 'Accept: application/octet-stream'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 10.0M  100 10.0M    0     0  20.3M      0 --:--:-- --:--:-- --:--:-- 20.3M
james@tamale|t.main:~/bin$
::: file 63040612
63040612: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=ca974c7e079f34babf7575cd2a2e5fd8022e662b, for GNU/Linux 3.2.0, with debug_info, not stripped
```

**Comment by jaemk** [comment]
It'd be helpful if you could modify your code to use the example code linked to in https://github.com/jaemk/self_update/issues/81#issuecomment-1019003981, modify the final `Move::from_source` to write to some other file, and then check if that file has the full contents. My guess is it's something to do with replacing the current binary, but I'm unable to reproduce it myself.

**Comment by andrew121410** [comment]
This works 

![Screenshot 2022-04-19 214414](https://user-images.githubusercontent.com/12376951/164130485-e6d2c894-deaa-482b-b0d1-253130c8ba94.png)

![Screenshot 2022-04-19 214903](https://user-images.githubusercontent.com/12376951/164130584-a4ec5808-960f-44b5-857b-609fe795ab24.png)

**Comment by andrew121410** [comment]
Now this actually works. I'm beyond confused

![Screenshot 2022-04-19 215816](https://user-images.githubusercontent.com/12376951/164131651-186f240d-5d6a-4d36-ac9a-b07b408403b0.png)

**Comment by jaemk** [comment]
That's... Interesting. And if you replace the contents of that update function with a call to the GitHub updater, then it spits out an empty file again?

**Comment by andrew121410** [comment]
Yes replacing it with this still replaces it with an empty binary

![A4592734-24DB-4052-8354-3DCFB1473A9E](https://user-images.githubusercontent.com/12376951/164149648-ffb5c586-ad99-4af0-b711-a9aa72ff50e0.jpeg)

I just realized it’s probably not working correctly because I’m not utilizing self_update correctly in update.rs it calls this

![CAF10A1B-9464-4281-B821-A2CAA13882A0](https://user-images.githubusercontent.com/12376951/164149941-61692852-6255-4110-9226-7c55eb363092.jpeg)

My only asset is just a binary not a archive like a .zip etc

![412CAEB0-7682-46B9-859F-8A0905A6C2EC](https://user-images.githubusercontent.com/12376951/164150036-ef56e4bc-5309-42d5-8e04-d6021f2acccd.jpeg)

**Comment by jaemk** [comment]
Ah, yeah that'll do it! Glad you were able to get to the bottom of it. Maybe it was the same thing affecting @Nickhoyer
